go-controller/pkg/ovn/policy.go

```diff
@@ -222,33 +223,67 @@ func deleteACLPortGroup(portGroupName, direction, priority, match, action string
 	return nil
 }
 
-func addToPortGroup(ovnNBClient goovn.Client, portGroupName string, portInfo *lpInfo) error {
-	cmd, err := ovnNBClient.PortGroupAddPort(portGroupName, portInfo.uuid)
-	if err == nil {
-		if err = ovnNBClient.Execute(cmd); err != nil {
-			return fmt.Errorf("execute error adding port %s to port group %s (%v)", portInfo.name, portGroupName, err)
+func addToPortGroup(ovnNBClient goovn.Client, portGroupName string, ports ...*lpInfo) error {
+	cmds := make([]*goovn.OvnCommand, 0, len(ports))
+	for _, portInfo := range ports {
+		cmd, err := ovnNBClient.PortGroupAddPort(portGroupName, portInfo.uuid)
+		if err != nil {
+			return fmt.Errorf("error preparing adding port %v to port group %s (%v)", portInfo.name, portGroupName, err)
 		}
-	} else if err != goovn.ErrorExist {
-		// Ignore goovn.ErrorExist to implement port "--may-exist" behavior
-		// i.e., if the port is already there, it's okay.
-		return fmt.Errorf("error adding port %s to port group %s (%v)", portInfo.name, portGroupName, err)
+		cmds = append(cmds, cmd)
+	}
+
+	if err := ovnNBClient.Execute(cmds...); err != nil {
+		names := []string{}
+		for _, portInfo := range ports {
+			names = append(names, portInfo.name)
+		}
+		return fmt.Errorf("error committing adding ports (%v) to port group %s (%v)", strings.Join(names, ","), portGroupName, err)
 	}
 	return nil
 }
 
-func deleteFromPortGroup(ovnNBClient goovn.Client, portGroupName string, portInfo *lpInfo) error {
-	cmd, err := ovnNBClient.PortGroupRemovePort(portGroupName, portInfo.uuid)
-	if err == nil {
-		if err = ovnNBClient.Execute(cmd); err != nil {
-			return fmt.Errorf("execute error removing port %s from port group %s (%v)", portInfo.name, portGroupName, err)
+func deleteFromPortGroup(ovnNBClient goovn.Client, portGroupName string, ports ...*lpInfo) error {
+	cmds := make([]*goovn.OvnCommand, 0, len(ports))
+	for _, portInfo := range ports {
+		cmd, err := ovnNBClient.PortGroupRemovePort(portGroupName, portInfo.uuid)
+		if err != nil {
+			// PortGroup already deleted...
+			if err == goovn.ErrorNotFound {
+				return nil
+			} else {
+				return fmt.Errorf("error preparing removing port %v from port group %s (%v)", portInfo.name, portGroupName, err)
```

Comment:
if we get stuck in a situation where one port has a problem, we are returning out of this function immediately, rather than trying to delete what we can and returning an error for the ports that failed. In ovn-kube delete is usually best effort, and in most of the cases where deleteFromPortGroup is called, the error is simply logged. Would it make more sense to try to remove everything we can, and still return an error?

Reply:
So, there are only two cases that return an error when preparing the command:
1 - the PortGroup itself doesn't exist (or the db is down) - so clearly we can return
2 - the UUID can't be parsed as a valid UUID - so something is very broken.

So, we don't really need to worry about the case where a single prepare fails.